Add README

tiwanari opened the pull request:
I'm writing README.md for drive. I want to discuss what we should include (or exclude) in it.

tiwanari committed:
refactor views

tiwanari committed:
Refactor titles (#29)

* refactor titles

* remove useless override

* remove useless NxtSettingActivity

tiwanari committed:
Replace string keys in code with string keys in xml (#30)

* replace string keys in code with string keys in xml

* revert private members

tiwanari committed:
Refactor execution package (#34)

* refactor execution package

* fixed comments on an Activity.

* update the appearance of selection blocks

* encupsalate ExecutionCondition and fix blocks based on the change

* [WIP] update codes based on comments

* update classes based on our discussion

* update based on comments

* add a conversion from Non-RandomAccess List to ArrayList

tiwanari committed:
Remove a useless conversion (#39)

* Remove a useless conversion

* Revert "Remove a useless conversion"

This reverts commit 35e8244d864d1ab2ecab483554bcedd34f3cd380.

* update based on comments

tiwanari committed:
Add README

Myusak committed:
Refactor Blocks (#35)

* Refactor Blocks

Following changes are made in this commit:
* Rename classes
    As we discussed offline

* Extract common functions among block classes
    Changed NumberTextHolder interface for better encapsulation

* Change return value type of BlockBase#getKind
    The previous code uses Class class, which does not represent actual
    use case (e.g., ForwardSecBlock#getKind returns SequenceBlock.class).
    Replaced Class<? extends Block> type with Enum.

* Remove locale-dependent code on parsing numbers
    The previous code relies on the Locale.US no matter what locale
    this app runs on.
    This commit fixes the locale dependent code by extracting the parsing
    logic.

* Fix the unspoken rule that using scaled values
  for communicating between blocks and number pickers.
    The previous code scaled values so that the value has no radix
    point, which was confusing code.
    Fixed by replacing the type with BigDecimal and explicitly use the raw
    value.
    In accordance with it, changed the type of "number"
    in ProgramDataSpec for saving without data loss

* Remove code in Unit class for handling quantities, which was ugly and cannot be used widely.
    Changed second string ("second/s" -> just "s")

* Fix PR according to the comments on #35

* remove UNDO in BlockFactory.java
* move FOREVER_WHILE_OFFSET in RepetitionBlock.java to ExecutionCondition.java
* add Javadocs on NumberTextDelegate
* change the signature of BlockSpaceManagerBase#placeBlock (ArrayList to List)
* rename OnTouchNumTextListener to OnTouchNumberTextListener
* remove OnNumberTextListener#context and rename holder to mHolder
* rename Unit#getUnitString to decorateValue and introduce new class NumberUtil

* Add and fix tests

for the following classes:
* NumberTextViewDelegate
* Unit
* NumberUtil
* ParseUtil

And cradle.build in the root directory is modified to add the dexmaker libraries
which are necessary for mockito

* Remove superfluous line-breaks

* Add constructor guards for util classes

mandaiy committed:
Fix formatting error in Unit.java

This commit resolves an error which occurred
at the moment of generating an NTimes block.
The cause of the bug was casting illegally
from BigDecimal to Integer class in Unit.java.

tiwanari committed:
Allow to choose the firmware of nxt (#38)

* Allow to choose the firmware of nxt

* add final to a member

* add ja/string for firmware setting

* fix the orientation of radio buttons in the firmware setting

Myusak committed:
Enable TravisCI (#44)

This pull-request introduces continuous integration on TravisCI.
This also resolves and closes #37.
With this PR, you can ensure that this project can be built without error.

This PR also removes the test in androidTest which uses Activity.
The test failed due to the package problem, which should be resolved in future.
The cause of the failure was Gradle cannot determine correct packages during testing,
as @tiwanari mentioned in #37.
(The test seemed awkward compared to ones other developers created.
So anyway the test had to be modified.)

Finally the failure of NumberTextViewDelegate test was resolved, which was caused by
the writer's fault and lack of understanding.

tiwanari commented:
It's time to make this PR finished and this repository open, isn't it? :tada:

tiwanari commented:
Why Travis says 'fails'?

mandaiy commented:
Do you have any solid reasons you have merged this instead of rebasing?
Rebasing is preferable in a case where the mainline has not changed the same files.

And in my opinion, I'm sorry that, we should not open (make announcement) this repository
unless it has our pile robot features.

tiwanari commented:
Thanks @myusak ,

> Do you have any solid reasons you have merged this instead of rebasing?

No, actually. I just want to discuss README by using a PR not an issue. I agree with you and will rebase it after getting approvals. Is there any comment on README itself?

> And in my opinion, I'm sorry that, we should not open (make announcement) ...

Oh, I think it is also okay to wait for the feature of our original robots but opening it just with the nxt feature is also okay and valuable not only for others but also for us because anyone can check our codes and may find some bugs. What do you think?

Anyway, regardless of its decision, I think we have to consider whether there is any problem if we open NxtDrive alone. If no, we can do the same refactorings to other product flavors and - if it is okay - open it.

mandaiy commented on README.md:
```diff
@@ -1,2 +1,75 @@
+=====
+[![Build Status](https://travis-ci.org/PileProject/drive.svg?branch=master)](https://travis-ci.org/PileProject/drive)
+
+Android applications provide children to learn programming
```
英文として, 「（世間一般の ） Android アプリケーションは」という主語になる気がします。
また "provide sth to" は不定詞をとれないような。

mandaiy commented on README.md:
```diff
@@ -1,2 +1,75 @@
+[![Build Status](https://travis-ci.org/PileProject/drive.svg?branch=master)](https://travis-ci.org/PileProject/drive)
+
+Android applications provide children to learn programming
+with a visual programming and a real robot.
```
"visual programming" には冠詞いらない気が

mandaiy commented on README.md:
```diff
@@ -1,2 +1,75 @@
+* 1.0.0
+    * First: Release NxtDrive.
+
+## Meta
```
Don't we need our website?

mandaiy reviewed (approved):
Added comments.
But I'm not good at English, so if you feel my comments are garbage, you can go ahead.

mandaiy commented:
Hmm, ok, I understood your opinion.
But I feel our code still needs refactoring.
So I cannot agree with `It's time to make this PR finished and this repository open`.

tiwanari commented:
Thanks, 

> But I feel our code still needs refactoring.

This is what I wanted to know 👍 So, what features we should refactor? We should make a milestone.

mandaiy commented:
At least we should resolve:
- #43 
- #40 
- #24 
- making sure that all files have copyright notice
- making sure that all files are applied our formatter
- (#3 seems to be resolved)

The new github feature, [project](https://github.com/PileProject/drive/projects) could boost us for release, you should try it.

In addition to them I hope to resolve:,
- #45
- #33 
- bump [SquiDB](https://github.com/yahoo/squidb) version up to 3.x 

For further improvements, I really want the feature of compiling blocks before execution.
That is essentially to generate an `ArrayList`, each of whose element is a pair of block and next address,
that array might be like:

```
+------+-----+-------+---------+-----+-----+-------+-------------+------+
|      |  0  |   1   |    2    |  3  |  4  |   5   |      6      |  7   |
+------+-----+-------+---------+-----+-----+-------+-------------+------+
| Kind | Seq | While | If      | Seq | Seq | EndIf | EndWhile    | TERM |
|      | 1   | 2     | True:3  | 5   | 5   | 6     | overzero:2  |      |
| Info |     |       | False:4 |     |     |       | otherwise:7 |      |
|      |     |       |         |     |     |       | counter     |      |
+------+-----+-------+---------+-----+-----+-------+-------------+------+
```

One execution is done by sequentially accessing the array from 0-th index, and the next block address can be fetched by the associated information.
I believe that kind of process is superior to the current solution.
Obviously this needs more consideration, so you can ignore it now (If I can come up with a neat solution, I will make an issue or a PR).

mandaiy commented:
And do you have a concrete deadline?

tiwanari commented:
Hi, thanks for listing up stuff 🙇 
I made a project for [`Release`](https://github.com/PileProject/drive/projects/1). 

> And do you have a concrete deadline?

We have a plan to update our website to advertise our products and kick it into high gear from around Dec. 20. So, I think it is good to make this repository open at least by the end of this year.

tiwanari committed:
Hotfix/execution bugs (#46)

* fix bug on ExectutionThread and BlockProgramLogic

* add a notice to show the error of connection

* fix a bug on ExecutionActivityBase

* remove is_connected data. it was uglily scattered in many places to notify an Android device whether it is already connected to a robot. make it worse, it caused some bugs

* add null checks for sensors

* remove useless Log.d

* add BluetoothCommunicatorTest

* add BlockProgramLogicTest

* also add mokito for unit testing

* add NxtControllerTest and fix bugs

* remove useless conversion

* update tests based on review comments

* change interfaces of BluetoothCommunicator / NxtController

* fix a little based on an interface change

* add the latest dexmaker by using JitPack.io

* fix a bug on a test

* avoid instantiate BluetoothDevice by specifying @NonNull

* comment out a test

* fix BluetoothCommunicator to throw NPE for null device

* comment out all tests related to PowerMock

* follow review comments

* revert build.gradle changes related to dexmaker

* add  modifier

tiwanari committed:
Upgrade SquiDB from 2.0.2 to 3.1.3 (#48)

tiwanari committed:
Do check/connectedCheck at once (#50)

tiwanari committed:
Hotfix/setting bugs (#51)

* fix strings to remove build warnings

* fix a bug on swapping ports

* change the string resource for second

* fix modifiers on members

* fix the collapsed threshold dialog by specifying the width of it

Myusak committed:
Introduce Dagger2 (#52)

This PR closes #31 and #45.

Following changes are made:

Renamed Activities' name by removing suffix "Base"
Changed intent-creation strategy. An intent for an Activity is created by the Activity's static method
Added DI code for BlockProvider and MachineProvider
In addition to these changes, minor changes were made

Moved Block category constants to @BlockCategory annotation interface
Changed the way of invoking MachineBase.connect (Thread to RxJava) in ExecutionActivity
Refactored some code

Myusak committed:
Add RxObservableProgram for program execution (#55)

This commit removes ExecutionThread class.
Instead of the class, RxObservableProgram is added,
which is Rx version of the Thread class.

This commit also fixes lifecycle-related bugs of ProgressDialogFragment and
AlertDialogFragment

tiwanari committed:
Add Copyright Notice (#57)

* add icense-gradle-plugin

* add CopyrightNoticeFragment and a simple script (WIP)

* update LICENSE and refer it from gradle

* update the headers of Java files by using licenseFormat command

* refactor .travis.yml and change the android target version

* add tasks to clean up licenses

* rename license files

* replace the plugin to make licenses.html with cookpad's one

* remove scripts folder

* readd license-gradle-plugin

* update xml's licenses

* ignore *Test.java for license checking

* fix an error on app/src/main/res/layout/activity_setting.xml

* fix an error on app/src/main/res/layout/content_execution.xml

tiwanari committed:
Make sudo required in .travis.yml (#60)

tiwanari committed:
Apply the license format to tests (#58)

tiwanari committed:
Standardize controllers (#59)

This commit standardizes the MachineController to make it has the union of its children's methods.
We keep MachineController as it were and add 'CarControllerBase'
for car-formed machines such as the car form of Nxt / Ev3 and Pile robots,
and rename NxtController after it (NxtCarController).
By this change, the previous block classes and resources related to the car-formed machines
can be moved in 'main' variant and we decide to keep them in 'car' package for readability
(StopForSecBlock is not moved because it doesn't depend on the car-formed machines' method).

This commit also cleans string resources.

* TL;DR
** Background
We talked over the design of controllers again and again offline, and we had two options;

- Make controllers for each machine (Nxt, Ev3, Pile, etc.) so that they have their specific methods
(e.g., Pile Robot doesn't have ColorSensor, thus it doesn't have getColorSensorValue or something like that)
and keep the MachineController as it is.
- Make the common MachineController that has the union of methods of these controllers.

They actually have pros & cons and our opinions conflicted. But finally, at present, we decided to choose the latter.

** Pros & Cons of the former
*** Pros
- Straightforward way and thus readable. This is very important for developers.
- They can be easily tested.

*** Cons
- There will be many duplications of codes because these controllers use [our library](https://github.com/PileProject/drivecommand)'s
MachineBase that provides the same interface to control machines (so, we will have very similar codes in blocks like
[ForwardSecBlock](https://github.com/PileProject/drive/blob/develop/app/src/nxt/java/com/pileproject/drive/programming/visual/block/sequence/ForwardSecBlock.java)
in each build variant) but we currently can't make the most of it because of the design.
- It doesn't follow the design of [our library](https://github.com/PileProject/drivecommand). But we can fix that of our library easily.

** Pros & Cons of the latter
*** Pros
- It follows the design of [our library](https://github.com/PileProject/drivecommand).
- It commonizes the blocks through MachineController interface. Thus, it reduces the duplicated codes not only *.java but also *.xml.
- Easy to maintain because the codes are unified.

*** Cons
- It is difficult to test each variant.
- Not straightforward way.
- Users can't know what kinds of methods each controller has, and make it worse,
calling wrong methods lead to RuntimeError not CompileError.
- Poor extensibility. We should add/remove a method to/from MachineController itself if new machines with new methods.
So, maybe we should make an another layer for other machines something
like MachineController -> CarController, DroneController (i.e., not use the MachineController as the current NxtController's parent).

** Why we chose the latter
This was a very difficult problem for us but we took the pros of the latter.
Maintaining duplicated codes *.java and *.xml is very annoying
and we can easily change our mind when a new type of machine comes
if we keep the unified codes clean.

tiwanari committed:
Add Legal Notice (#61)

This commit introduces the legal notice of our apps and also merges the general codes for showing plain HTML in a fragment.

* How to update it
1. modify `app/src/main/assets/legal_notice_template.md`

2. use [pandoc](https://www.google.co.jp/search?q=pandoc&oq=pandoc&aqs=chrome..69i57j69i65l2j0l3.890j0j1&sourceid=chrome&ie=UTF-8) to convert the above markdown to a html.

```
pandoc legal_notice_template.md
```

Then, you'll get the converted result.

3. copy the result and paste it into the `article` element of `app/src/main/assets/legal_notice.html`


* NOTE
I used [sindresorhus/github-markdown-css: The minimal amount of CSS to replicate the GitHub Markdown style]( https://goo.gl/kZ7YZY ) for the pages' css.

tiwanari committed:
Clean up comments (#62)

Solves #24. This is a PR to clean up comments/javadocs. This may be related to #56.

# Background
Before moving on to adding other variants (ev3/pile), I think we should fix some comments/javadocs because we should do the same changes to the other variants if we implement them. At present, some javadocs have errors with their grammar.


# Goal
The goal is not enriching but cleaning comments up. Of course, I add some information but they may be not enough. 


# Note
I add a plugin to generate javadocs automatically. It introduces some tasks like `generateNxtDebugJavadoc`. Running these tasks causes some warnings because of the reference errors of auto-generated classes but we can see the complete Javadocs in `app/javaDoc`. I believe that the document is really helpful not only for others but also for us.

However, I'm sorry that we should care about the tasks so that they shouldn't have errors related to auto-generated classes. But please don't worry about that. I will check them when you submit a new PR.

tiwanari commented:
Hi @myusak @amiq11 

I came up with adding some screenshots into README and I'll prepare them. 
Do you have any more suggestions?:)

tiwanari commented:
Hi @makotoshimazu @myusak 

Sorry for sending many PRs at one time but please check this PR, too.
This PR has not many changes. 
I added some screenshots and updated README according to the images.

If you give me an approval, I will rebase after merging the current `develop` to `master`.
Thank you in advance.

tiwanari commented:
Thank you for your check, @myusak 👍 

I will rebase it if @makotoshimazu doesn't refuse this change by next Saturday.

tiwanari commented:
OK, so let me merge this:)

tiwanari committed:
fix javadocs

tiwanari commented:
Sorry, I made the wrong things on this PR... but [this commit](https://github.com/PileProject/drive/commit/71503bed2056d8b5faf578ebcd964ec402d6ceda) is what I did on this PR. Please ignore the current PR's changes.
